### PR TITLE
Converting Testing To Allow Either Codeception Based Testing or PHPUnit

### PR DESCRIPTION
**Pull request by Fenikkusu**
Hello!

* Type: new feature

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/incubator/blob/master/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

There's a lot going on here, but it's actually fairly simple. Currently, the Phalcon Test Classes are all built specifically for PHPUnit. Even if you are using Codeception to run your tests, your limited to using the PHPUnit setup and you loose some of the extra benefits of using the Codeception base instead (IE: _before shorthand, Test Helpers, DB Methods, etc). This PR is an attempt to bridge that gap while also maintaining some backwards compatibility. In order to accomplish this, a number of changes were made:

1) Each of the test cases were converted to traits and moved to Library\Test\Traits.
2) Slight changes were made to the new traits to make the platform independent (Renamed setUp/tearDown to setUpPhalcon/tearDownPhalcon respectively).
3) New classes were created in Library\Test\PHPUnit that use the new traits and call the setUp/tearDown functions appropriately.
4) New classes were created in Library\Test\Codeception that also use the new traits and call the setUp/tearDown functions appropriately.
5) Placeholder classes replaced the original TestCases that just extend the appropriate PHPUnit edition of the class.

I've added a number of Unit Tests in place to help ensure this logic is working correctly and thus far I've had no issues. I'm going ahead and creating the PR so others can look over the tests while I run more direct tests to ensure everything is working as intended.

I should mention that while I've done my best to make these changes backwards compatible, there is one scenario where they will not be backwards compatible. If someone were to ever use an 'instanceof' command on the placeholder TestCases, then these may break. This seems highly unlikely, but I wanted to make sure to mention it. For example:

```
$testCase = new \Phalcon\Test\FunctionalTestCase(); //Abstract Class, Not Technically Possible
if ($testCase instance of \Phalcon\Test\UnitTestCase) {
  //This will never be reached now.
}
```

PS: Way to many 'Test's - UnitTestCase, UnitTestCase, UnitTestCase, & UnitTestCaseTest

### CODE REVIEW

**Comment by Fenikkusu** [comment]
Quick Update. I was able to install this on an existing project with PHPUnit and run the testing. Thus far I've seen no issues. I'm working on running it on codeception as well.

**Comment by sergeyklay** [comment]
@Fenikkusu It would be great if you could test the Incubator by Incubator :)

**Comment by Fenikkusu** [comment]
@sergeyklay ...excuse me while I have a stupid moment...but huh?

On another note, do you know if there is anything I can do about the failing tests (I honestly haven't looked into).

**Comment by Fenikkusu** [comment]
May I take to assume that you wish me to update all the tests to use the new Codeception Testing Classes? IE: Update Phalcon\Tests\Queue\Beanstalk\ExtendedTest to extend the new Phalcon\Test\Test\Codeception\UnitTestCaseTest?

Keeping the whole Codeception concept going.

**Comment by sergeyklay** [comment]
I mean you can use these classes as base classes for some in `tests/` folder. Not necessarily for all tests, but some simple tests could use them as parent.

About failed test: something went wrong here: 
https://github.com/phalcon/incubator/blob/e35352e27fb25521686da4acc82bbffa208c206b/Library/Phalcon/Validation/Validator/Db/Uniqueness.php#L85-L87

**Comment by Fenikkusu** [comment]
@sergeyklay I was able to get the DB Error Fixed. Not sure if it's the correct method or not, but its worth mentioning. I know that Phalcon\Di calls setDefault in the construct. However, I've actually had issues on a number of occasions where I needed to explicitly call setDefault outside of the construct. This is one of those times. I've yet to understand why this is, but I've also not dived into it any.

I'll work on the updates when time permits.

File: Library/Phalcon/Test/Codeception/UnitTestCase.php [hunk 1]
```diff
@@ -21,11 +21,17 @@
 use Codeception\Test\Unit;
```
Comment (sergeyklay) [inline Library/Phalcon/Test/Codeception/UnitTestCase.php]: To use `Codeception\*` we'll need to move Codeception from `require-dev` to `require`

Comment (Fenikkusu) [reply Library/Phalcon/Test/Codeception/UnitTestCase.php]: The only problem I have with that is that PHPUnit is and has been a require-dev for a while anyways. Plus I don't really see the need. They are both only needed when running the incubator's testing and the rest of the time it's up to the user to install it. Additionally such a change would force codeception to always be installed, even if a person does --no-dev, would it not? I think that would upset people. I myself wouldn't want codeception in a production environment.

**Comment by Fenikkusu** [comment]
Done

**Comment by Fenikkusu** [comment]
Hmm...PHPCS Is Scanning Vendor Folder...

**Comment by Fenikkusu** [comment]
Scratch That. Symphony/Latest is using PHP7.1 syntax now...I'll let someone else handle that.

**Comment by Fenikkusu** [comment]
Rebased Again to include passing CI updates.

**Comment by sergeyklay** [comment]
@Fenikkusu Could you please change target branch for this PR to 3.3.x

**Comment by Fenikkusu** [comment]
@sergeyklay Done

**Comment by sergeyklay** [comment]
Thank you

**Pull request merged**